dvc/repo/diff.py

```diff
@@ -101,17 +118,29 @@ def _to_checksum(output):
     for stage in repo.stages:
         for output in stage.outs:
             if _exists(output):
-                yield _to_path(output), _to_checksum(output)
-                if output.is_dir_checksum:
-                    yield from _dir_output_paths(repo_tree, output)
+                if targets is None or any(
+                    output.path_info == target for target in targets
```

Comment:
Seems like this should be `output.path_info.isin_or_eq(target)` and not `==`?

If I have
```
.
└── dir
    └── output
```
and `targets=["dir"]`, this won't match `output`, since `output.path_info` will be `dir/output`

Reply:
So that case will be covered by the `elif` path below:
```python
                elif output.is_dir_checksum and any(
                     target.isin(output.path_info) for target in targets
                 ):
                     yield from _dir_output_paths(repo_tree, output, targets)
```

I decided to have two paths because for the `== and output.is_dir_checksum` path `_dir_output_paths` can be called without any `targets` to filter because we want all files in that folder for the diff.

Reply:
What about the case where `output` is a file output and not a directory output (where `output.is_dir_checksum` is false)?

Reply:
Please disregard my previous comment, for now.

Reply:
re: https://github.com/iterative/dvc/pull/4938#discussion_r535952997

I had missed the `dvc add dir/output` vs `dvc add dir` case in my original tests in https://github.com/iterative/dvc/pull/4938/commits/4f3a99a85ab98a81b8af9279fab768bb452bcc78. With your explanation now understand some of the simplifications made in https://github.com/iterative/dvc/pull/4938/commits/395167309234a0fb80cd66a416d3b52cc43f2e1e do not cover this any more.

I've added this scenarios to the tests in https://github.com/iterative/dvc/pull/4938/commits/4b4f04b49099ae24f48d248a7a787540efd36ab8 along with going back to the logic from https://github.com/iterative/dvc/pull/4938/commits/2b3e888758f08ea2464eede67481c62c3a8d346d but using `any` like you suggested.